src/maxflow.rs

```diff
@@ -1 +1,251 @@
+use crate::internal_queue::SimpleQueue;
+use std::cmp::min;
+use std::iter;
+use std::ops::{Add, AddAssign, Sub, SubAssign};
 
+// Maybe it should be in a separate module
+pub trait MfCapacity:
+    Copy + Ord + Add<Output = Self> + AddAssign + Sub<Output = Self> + SubAssign
+{
+    fn zero() -> Self;
+    fn max_value() -> Self;
+}
+
+impl MfCapacity for i32 {
+    fn zero() -> Self {
+        0
+    }
+    fn max_value() -> Self {
+        std::i32::MAX
+    }
+}
+
+impl MfCapacity for i64 {
+    fn zero() -> Self {
+        0
+    }
+    fn max_value() -> Self {
+        std::i64::MAX
+    }
+}
+
+impl MfCapacity for u32 {
+    fn zero() -> Self {
+        0
+    }
+    fn max_value() -> Self {
+        std::u32::MAX
+    }
+}
+
+impl MfCapacity for u64 {
+    fn zero() -> Self {
+        0
+    }
+    fn max_value() -> Self {
+        std::u64::MAX
+    }
+}
+
+impl<Cap> MfGraph<Cap>
+where
+    Cap: MfCapacity,
+{
+    pub fn new(n: usize) -> MfGraph<Cap> {
+        MfGraph {
+            _n: n,
+            pos: Vec::new(),
+            g: iter::repeat_with(Vec::new).take(n).collect(),
+        }
+    }
+
+    pub fn add_edge(&mut self, from: usize, to: usize, cap: Cap) -> usize {
+        assert!(from < self._n);
+        assert!(to < self._n);
+        assert!(Cap::zero() < cap);
+        let m = self.pos.len();
+        self.pos.push((from, self.g[from].len()));
+        let rev = self.g[to].len();
+        self.g[from].push(_Edge { to, rev, cap });
+        let rev = self.g[from].len() - 1;
+        self.g[to].push(_Edge {
+            to: from,
+            rev,
+            cap: Cap::zero(),
+        });
+        m
+    }
+}
+
+struct Edge<Cap> {
+    from: usize,
+    to: usize,
+    cap: Cap,
+    flow: Cap,
+}
+
+impl<Cap> MfGraph<Cap>
+where
+    Cap: MfCapacity,
+{
+    fn get_edge(&self, i: usize) -> Edge<Cap> {
+        let m = self.pos.len();
+        assert!(i < m);
+        let _e = &self.g[self.pos[i].0][self.pos[i].1];
+        let _re = &self.g[_e.to][_e.rev];
+        Edge {
+            from: self.pos[i].0,
+            to: _e.to,
+            cap: _e.cap + _re.cap,
+            flow: _re.cap,
+        }
+    }
+    fn edges(&self) -> Vec<Edge<Cap>> {
+        let m = self.pos.len();
+        (0..m).map(|i| self.get_edge(i)).collect()
+    }
+    fn change_edge(&mut self, i: usize, new_cap: Cap, new_flow: Cap) {
+        let m = self.pos.len();
+        assert!(i < m);
+        assert!(Cap::zero() < new_flow && new_flow <= new_cap);
+        let (to, rev) = {
+            let _e = &mut self.g[self.pos[i].0][self.pos[i].1];
+            _e.cap = new_cap - new_flow;
+            (_e.to, _e.rev)
+        };
+        let _re = &mut self.g[to][rev];
+        _re.cap = new_flow;
+    }
+
+    pub fn flow(&mut self, s: usize, t: usize) -> Cap {
+        self.flow_with_capacity(s, t, Cap::max_value())
+    }
+    pub fn flow_with_capacity(&mut self, s: usize, t: usize, flow_limit: Cap) -> Cap {
```

Comment:
What is the reason of this naming? It make sense to name this a `capacity` in some specific context, but here probably doesn't. Is there any reference that this is called capacity?
Maybe `flow_with_limit`, `flow_with_value_limit` or something like that?

Reply:
I have an image of adding new edge from virtual vertex `s'` to vertex `s` with capacity `flow_limit`.  But I agree that `flow_with_limit` may be more intuitive.  I want to ask others' opinion for this point.